service/controller/key/generate/ami.go

```diff
@@ -0,0 +1,186 @@
+//go:generate go run ami.go
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/semver/v3"
+	"golang.org/x/net/html"
+
+	"github.com/giantswarm/aws-operator/service/controller/key"
+)
+
+const (
+	primaryDomain     = "flatcar-linux.net"
+	chinaDomain       = "flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn"
+	minimumVersion    = "2191.5.0"
+	channel           = "stable"
+	arch              = "amd64-usr"
+	generatedFilename = "ami.go"
+	generatedPackage  = "key"
+	generatedTemplate = `// NOTE: This file is generated. Do not edit.
+package {{ .Package }}
+
+import "encoding/json"
+
+var amiJSON = []byte({{ .AMIInfoString }})
+var amiInfo = map[string]map[string]string{}
+
+func init() {
+	err := json.Unmarshal(amiJSON, &amiInfo)
+	if err != nil {
+		panic(err)
+	}
+}
+`
+)
+
+type sourceFileTemplateData struct {
+	AMIInfoString string
+	Package       string
+}
+
+func scrapeVersions(source io.Reader) ([]string, error) {
+	z := html.NewTokenizer(source)
+	var versions []string
+	for {
+		tt := z.Next()
+		switch {
+		case tt == html.ErrorToken:
+			return versions, nil
+		case tt == html.StartTagToken:
+			t := z.Token()
+			if t.Data != "a" {
+				continue
+			}
+			for _, attr := range t.Attr {
+				if attr.Key != "href" {
+					continue
+				}
+				// Versions to extract look like href="./123.4.5/" or href="123.4.5"
+				// so we trim off suffix and prefix if they exist and then ensure this
+				// is a valid semver version.
+				href := strings.TrimSuffix(attr.Val, "/")
+				if strings.HasPrefix(href, "./") {
+					href = strings.TrimPrefix(href, "./")
+				}
+				if _, err := semver.NewVersion(href); err != nil {
+					break // href is invalid, no need to look at other attrs
+				}
+				versions = append(versions, href)
+			}
+		}
+	}
+}
+
+func scrapeVersionAMIs(source io.Reader) (map[string]string, error) {
+	body, err := ioutil.ReadAll(source)
+	if err != nil {
+		return nil, err
+	}
+
+	var list key.AMIInfoList
+	err = json.Unmarshal(body, &list)
+	if err != nil {
+		return nil, err
+	}
+
+	result := map[string]string{}
+	for _, region := range list.AMIs {
+		result[region.Name] = region.HVM
+	}
+
+	return result, nil
+}
+
+func main() {
```

Comment:
I am not sure how this here works. Who is executing it and when? Do we need docs for it so humans get it? Can we automate it? What about having something for this in `devctl`? I am not sure this is well put somewhere in the operator codebase. Why don't we just put it in the `installations` repo and have it safely there? That would maybe enable us to deactivate AMIs in published releases in case we ever have to, maybe for security reasons.

Reply:
Currently the information about AMIs is completely done "by hand" in a key file:
https://github.com/giantswarm/aws-operator/blob/47a27ffb9d90fcd1ecaedd1665139557765ab1c4/service/controller/key/common.go#L300-L323

The automation Thomas here introduced let's us generate that and create a commit for it. It is not a final solution or a perfect solution - it does however go into the direction of having one and is a definite improvement of the current status.

Putting this info into the installations repository would also introduce challenges / problems. We will think about this in the future more and improve here as this is a problem we are still facing. 

I would like to go forward here as this is an improvement even though it might not be perfect. 

Proper usage documentation in the readme is a good point though and should be included cc @tfussell

Reply:
This is definitely a good thing and I did not mean to talk it down. I just think it is the wrong place to put it. What is the problem putting it into the `installations` repo? I would like to understand this a little bit better.

Reply:
Putting it into the `installations` repo requires a lot of thought:
- AMIs are tied to the region therefore the combination of installation and region would give us one small set of actually useful AMIs
- The generation in `installations` works in a defaulting manner - with the current workflow we would therefore still have to put all AMIs on all installations (unless we rework how generation works in installations)
- The usage of AMIs in `installations` implies more investment into the draughtsman workflow as draughtsman applies those values. This could easily be blocked as we do not want to invest more into this draughtsman workflow.
- We do gain the ability to change AMIs after the fact through configuration but this comes with the risk of misconfiguration causing old clusters to suddenly use incompatible AMIs

So there is a whole discussion to be had here in my opinion. Team Ludacris will bring this discussion up in the future as this generation here is also blocking us:
We want to be able to change the OS version with a single commit in `releases` but would need to generate AMIs first. We only have this issue on AWS but want to focus on it later on.

Reply:
We talked and came to the conclusion that we are happier with a move of this generation script to `devctl` for now.

I will provide a PR.

Reply:
This is removed now.